Include/object.h

```diff
@@ -694,21 +700,38 @@ PyAPI_FUNC(void) _PyTrash_thread_destroy_chain(void);
 
 #define PyTrash_UNWIND_LEVEL 50
 
-#define Py_TRASHCAN_SAFE_BEGIN(op) \
+#define Py_TRASHCAN_BEGIN_CONDITION(op, cond) \
     do { \
-        PyThreadState *_tstate = PyThreadState_GET(); \
-        if (_tstate->trash_delete_nesting < PyTrash_UNWIND_LEVEL) { \
-            ++_tstate->trash_delete_nesting;
-            /* The body of the deallocator is here. */
-#define Py_TRASHCAN_SAFE_END(op) \
+        PyThreadState *_tstate = NULL; \
+        /* If "cond" is false, then _tstate remains NULL and the deallocator \
+         * is run normally without involving the trashcan */ \
+        if (cond) { \
+            _tstate = PyThreadState_GET(); \
+            if (_tstate->trash_delete_nesting >= PyTrash_UNWIND_LEVEL) { \
+                /* Store the object (to be deallocated later) and jump past \
+                 * Py_TRASHCAN_END, skipping the body of the deallocator */ \
+                _PyTrash_thread_deposit_object(_PyObject_CAST(op)); \
+                break; \
+            } \
+            ++_tstate->trash_delete_nesting; \
+        } \
+        /* The body of the deallocator is here. */
+#define Py_TRASHCAN_END \
```

Comment:
Out of consistency and for future proofness, I think this should still be `Py_TRASHCAN_END(op)`.

Reply:
> Out of consistency

If you really want to be consistent, it should be `Py_TRASHCAN_END(op, dealloc)`. But I don't see the need for such "consistency".

> and for future proofness

I don't think it's needed for future proofness either: if we ever require `op` in the `Py_TRASHCAN_END` macro, it would be better to store it in a temporary variable and use that instead. That also avoids potential errors where somebody plugs in different values for `op` in the BEGIN and END macros.

But in the end, this is just bikeshedding. So if you insist, I will add the useless `op` argument.

Reply:
Fair enough.